odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1160 @@
+            `OperatorLeftVectorMult`.
+        """
+        if other in self.domain.field:
+            if other == 0:
+                from odl.solvers.functional.default_functionals import (
+                    ZeroFunctional)
+                return ZeroFunctional(self.domain)
+            else:
+                return FunctionalLeftScalarMult(self, other)
+        else:
+            return super().__rmul__(other)
+
+    def __add__(self, other):
+        """Return ``self + other``.
+
+        If ``other`` is a `Functional`, this corresponds to
+
+            ``func1 + func2 == (x --> func1(x) + func2(x))``
+
+        If ``other`` is a scalar, this corresponds to adding a scalar to the
+        value of the functional:
+
+            ``func + scalar == (x --> func(x) + scalar)``
+
+        Parameters
+        ----------
+        other : `Functional` or scalar
+            `Functional`:
+            The `Functional.domain` and `Functional.range` of ``other``
+            must match this functional's  `Functional.domain` and
+            `Functional.range`.
+
+            scalar:
+            The scalar needs to be in this functional's `Functional.range`.
+
+        Returns
+        -------
+        add : `Functional`
+            Addition result.
+
+            If ``other`` is in ``Functional.range``, ``add`` is a
+            `FunctionalScalarSum`.
+
+            If ``other`` is a `Functional`, ``add`` is a `FunctionalSum`.
+
+        See Also
+        --------
+        Operator.__add__ : implementation of __add__ for Operators.
+        """
+        if other in self.domain.field:
+            return FunctionalScalarSum(self, other)
+        elif isinstance(other, Functional):
+            return FunctionalSum(self, other)
+        else:
+            return NotImplemented
+
+    # Since addition is commutative, right and left addition is the same
+    __radd__ = __add__
+
+    def __sub__(self, other):
+        """Return ``self - other``."""
+        return self.__add__(-1 * other)
+
+    @property
+    def grad_lipschitz(self):
+        """Lipschitz constant for the gradient of the functional"""
+        return self.__grad_lipschitz
+
+
+class FunctionalLeftScalarMult(Functional, OperatorLeftScalarMult):
+
+    """Scalar multiplication of functional from the left.
+
+    Given a functional ``f`` and a scalar ``scalar``, this represents the
+    functional
+
+        ``(scalar * f)(x) == scalar * f(x)``.
+    """
+
+    def __init__(self, func, scalar):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func : `Functional`
+            Functional to be scaled.
+        scalar : `float`
+            Number with which to scale the functional.
+        """
+        if not isinstance(func, Functional):
+            raise TypeError('`func` {!r} is not a `Functional` instance.'
+                            ''.format(func))
+
+        scalar = func.range.element(scalar)
+
+        if scalar == 0:
```

Comment:
Whatever default value you select for `grad_lipschitz` (`np.inf`, `None` or `np.nan`), this will return the zero functional, but with value for `grad_lipschitz` which won't be zero unless you do some tests elsewhere. Ergo: I will keep it as it is.